### DISCUSSION TITLE
Finetuing CodeLLaMA on LeanDojo

### DISCUSSION THREAD

**Discussion by Some-random** [lean-dojo/LeanDojo#5] [General]
In my ATP research, I've fine-tuned a LLAMA 7B model using lean_proof_recording and used lean-gym to interact with Lean. Now, I plan to switch to LeanDoJo for improved data collection and Lean interaction. Hence, I have two queries:

1  Which exact version of miniF2F is used in the LeanDoJo paper? Was Meta's bug-reduced version employed?
2  Which repository and commit ID can enable full mathlib3 tracing, as in lean_proof_recording, to assess my fine-tuned LLAMA's performance?

**Comment by yangky11** [comment]
I'm transferring this issue to Github Discussions (see https://github.com/lean-dojo/LeanDojo#questions-and-bugs)

Regarding the questions themselves, they can be found in the paper by searching for "commit" or in our data generation script: https://github.com/lean-dojo/LeanDojo/blob/main/scripts/generate-benchmark-lean3.ipynb

**Comment by wiio12** [comment]
Hi @Some-random, do you have any results on LLAMA 7B running with lean-dojo by now? It would be valuable to see how well the Reprover works on a larger language model~.

And Hi @yangky11, great work here! Are you planning on running the Reprover with larger language models?

**Reply by yangky11** [reply]
Hi, we're also curious to see but probably won't do it in the near future, due to constraints on computing resources. Other people are welcome to try it using ReProver's codebase.

**Reply by GanjinZero** [reply]
Also curious.

**Reply by wiio12** [reply]
Hi @yangky11 and @GanjinZero,

I've recently fine-tuned a CodeLlama 7b model using the [LeanDojo Benchmark](https://doi.org/10.5281/zenodo.8016385) dataset. The data was formatted in the GPT-f style, as follows:
```
GOAL <tactic state> PROOFSTEP <tactic>
```

For this, I utilized a training script adapted from [Metamath](https://github.com/meta-math/MetaMath). The CodeLlama 7b was fine-tuned using various learning rates: `5e-4` (as recommended in the LeanDojo paper), `1e-5`, `3e-5`, and `5e-5`, all with a batch size of 8. 

![leandojo](https://github.com/lean-dojo/LeanDojo/assets/19700931/d8e9ba1f-3c57-4a1c-a141-a59615189ed5)
![leandojo2](https://github.com/lean-dojo/LeanDojo/assets/19700931/6ce6bd24-5593-45c2-9e3c-7c462f8a1cbf)

The model with the `5e-5` learning rate achieved the lowest validation loss. However, the training showed a 'staircase' pattern in both validation loss and training loss.

For the model that yielded the lowest validation loss, I ran the BFS using the ReProver repository.  At each
search step, I use beam-search decoding to produce 8 tactic candidates. Using e=64 (as suggested in the original paper) will cause CUDA out-of-memory errors. The results below were obtained by testing on a subset of 100 theorem samples from the LeanDojo Benchmark test set (the first 100 problems in the test set):

| Model                          | Pass Rate @1 |
|--------------------------------|--------------|
| ReProver (without retrieval)   | 39%          |
| CodeLlama (fine-tuned)         | 41%          |

The improvement is marginal, considering the larger size and extensive pre-training of the CodeLlama model. I'm concerned that there might be issues with my training approach. Do you have any suggestions or insights that might help?

**Reply by GanjinZero** [reply]
Have you ever tried eval on minif2f performance? I am going to do some reproducing with larger size models.

**Reply by wiio12** [reply]
Not yet, I could do a full evaluation on miniF2F and LeanDojo Benchmark after I am sure that the model is well trained, I'm still unsure that the training process is ok, given the staircase loss and the marginal improvement.

**Reply by GanjinZero** [reply]
Staircase loss is very common during sft, it seems you have 25k sft data. Use 25k sft data * 3 epoch could be a good start point. You may have a lr decay to stabilize the training.

**Reply by wiio12** [reply]
There are 208759≈25k*bs8 examples in the dataset, and I set the training budget to 10 epochs. All the models reach minimal validation loss around the second epoch (I wonder if it is the same for 300m ReProver?). I already use the cosine learning rate decay, but I guess my training budget is too large, so the learning rate is not decaying fast enough, causing the staircase loss. I will try shrinking the training budget :)

**Reply by wiio12** [reply]
Hi @GanjinZero, I tried shrinking the training budget to 3 epochs, but the training curve and validation curve are nearly identical to previous one.

![leandojo3](https://github.com/lean-dojo/LeanDojo/assets/19700931/365e17a8-71a9-4c4e-badf-d972af01036e)

It seems like increasing the model size really don't adds much to performance? (AlphaGeometry only use 151m languge model)

**Reply by yangky11** [reply]
@albertqjiang is working on finetuning Mistral and may have insights.

**Reply by wiio12** [reply]
Recent update:

1. Reducing the training budget helps, even if the training loss and validation loss are nearly identical, the downstream pass@1 varies a lot: (The `tb` is the abbreviation for `training budget`)

| Model                          | Pass Rate @1 |
|--------------------------------|--------------|
| ReProver (without retrieval)   | 39%          |
| CodeLlama (lr=5e-5, bs=8, tb=10e)       | 41%          |
| CodeLlama (lr=5e-5, bs=32, tb=10e)    | 41%         |
| CodeLlama (lr=5e-5, bs=32, tb=3e)      | 46%        |
| CodeLlama (lr=5e-5, bs=128, tb=10e)   | 47%        |
| CodeLlama (Meta setting) (lr=5e-5, bs=100, tb=3.5e, weight_decay=0.1) | 46% |

![image](https://github.com/lean-dojo/LeanDojo/assets/19700931/4806c1ff-eb47-4e4c-b1a3-6a7864360516)


2. I just found this paper: [Temperature-scaled large language models for Lean proofstep prediction](https://mathai2023.github.io/papers/). The paper finetuned the CodeLlama with the LeanDojo Benchmark dataset.
<img width="1088" alt="image" src="https://github.com/lean-dojo/LeanDojo/assets/19700931/c39fb13e-7ee0-4086-8200-942d23162156">

**Reply by yangky11** [reply]
So the model becomes worse after being trained for too many epochs? That's very interesting!

**Reply by tangzhy** [reply]
Same observation on Mistral 7b:)  @albertqjiang @yangky11 wonder if you see that as well?

**Reply by albertqjiang** [reply]
yep you should not train too many epochs because it will eventually overfit.

there are things you can do to make that happen later though (the later the overfit, the better the perf I've found). This will come out in my next paper.
